src/sentry/tasks/on_demand_metrics.py

```diff
@@ -0,0 +1,273 @@
+
+    Spawns a series of child tasks :func:`process_widget_specs`, and limits them using
+    a stateful (cached) count + modulo to spread out the work over `total_batches` number of scheduled task runs.
+
+    It's safe, but not ideal if a particular child tasks fails to run in cases where the current state is reset since memcache is ephemeral.
+
+    ### Other
+    - The amount of work being done can be reduced once extraction is stateful in our db, which would allow us to
+        only iterate over widgets with currently enabled extraction.
+
+    # Ops
+    Killswitch option: `on_demand_metrics.check_widgets.enable`
+    - It is safe to turn off for a period of `_WIDGET_QUERY_CARDINALITY_TTL`, customer data should not be lost. After this time
+        high cardinality metrics may inadverently be stored for on-demand-extraction.
+    """
+    if not options.get("on_demand_metrics.check_widgets.enable"):
+        return
+
+    rollout = options.get("on_demand_metrics.check_widgets.rollout")
+    total_batches = options.get("on_demand_metrics.check_widgets.query.total_batches")
+    widgets_per_batch = options.get("on_demand_metrics.check_widgets.query.batch_size")
+
+    currently_processing_batch = _get_current_processing_batch(total_batches)
+
+    widget_query_ids = []
+    dashboard_widget_pre_rollout_count = 0
+    dashboard_widget_count = 0
+
+    for (widget_query_id,) in RangeQuerySetWrapper(
+        DashboardWidgetQuery.objects.filter(
+            widget__widget_type=DashboardWidgetTypes.DISCOVER, columns__len__gt=0
+        )
+        .exclude(conditions__contains="event.type:error")
+        .values_list("id"),
+        result_value_getter=lambda item: item[0],
+    ):
+        dashboard_widget_pre_rollout_count += 1
+
+        if ((widget_query_id % 1_000) / 1_000) > rollout:
+            # % rollout based on widget_id accurate to 0.1%
+            continue
+
+        batch_for_widget_query = _get_batch_for_widget_query(widget_query_id, total_batches)
+        if batch_for_widget_query != currently_processing_batch:
+            continue
+
+        widget_query_ids.append(widget_query_id)
+        dashboard_widget_count += 1
+
+        if len(widget_query_ids) >= widgets_per_batch:
+            process_widget_specs.delay(
+                widget_query_ids,
+            )
+            widget_query_ids = []
+
+    # Clean up any remaining widgets
+    if widget_query_ids:
+        process_widget_specs.delay(
+            widget_query_ids,
+        )
+
+    _set_currently_processing_batch(currently_processing_batch)
+    metrics.incr(
+        "task.on_demand_metrics.widgets.currently_processing_batch",
+        amount=currently_processing_batch,  # Helps correlate which batch is currently being processed with metrics
+        sample_rate=1.0,
+    )
+    metrics.incr(
+        "task.on_demand_metrics.widgets.pre_rollout.total",
+        amount=dashboard_widget_pre_rollout_count,
```

Comment:
Yep, I want to get all the counts here, we can drop some of these metrics for cost after rollout into GA since they aren't needed for long term support but helpful for insight during rollout.